src/learner.cc

```diff
@@ -207,10 +207,9 @@ void GenericParameter::ConfigureGpuId(bool require_gpu) {
       LOG(WARNING) << "No visible GPU is found, setting `gpu_id` to -1";
     }
     this->UpdateAllowUnknown(Args{{"gpu_id", std::to_string(kCpuId)}});
-  } else if (gpu_id != kCpuId && gpu_id >= n_gpus) {
-    LOG(WARNING) << "Only " << n_gpus
-                 << " GPUs are visible, setting `gpu_id` to " << gpu_id % n_gpus;
-    this->UpdateAllowUnknown(Args{{"gpu_id", std::to_string(gpu_id % n_gpus)}});
+  } else {
```

Comment:
Can we make the old behavior optional? If so then we can submit this PR to upstream xgboost and have a flag to switch between old behavior and your behavior.

Reply:
I had the same idea but wanted to fist check that this works, will add a parameter